nixos/modules/services/networking/seafile.nix

```diff
@@ -0,0 +1,272 @@
+        wantedBy = [ "seafile.target" ];
+        partOf = [ "seafile.target" ];
+        after = [ "network.target" "seaf-server.service" ];
+        requires = [ "seaf-server.service" ];
+        restartIfChanged = true;
+        restartTriggers = [ (seafileConf) ];
+        environment = {
+          PYTHONPATH =
+            "${pkgs.python3Packages.seahub}/thirdpart:${pkgs.python3Packages.seahub}:${penv}/${python.sitePackages}";
+          DJANGO_SETTINGS_MODULE = "seahub.settings";
+          CCNET_CONF_DIR = ccnetDir;
+          SEAFILE_CONF_DIR = dataDir;
+          SEAFILE_CENTRAL_CONF_DIR = centralConfigDir;
+          SEAFILE_RPC_PIPE_PATH = "/run/seafile";
+        };
+        serviceConfig = {
+          User = "seafile";
+          Group = "seafile";
+          DynamicUser = true;
+          ProtectHome = true;
+          PrivateUsers = true;
+          PrivateDevices = true;
+          ProtectClock = true;
+          ProtectHostname = true;
+          ProtectProc = "invisible";
+          ProtectKernelModules = true;
+          ProtectKernelTunables = true;
+          ProtectKernelLogs = true;
+          ProtectControlGroups = true;
+          RestrictNamespaces = true;
+          RestrictAddressFamilies = [ "AF_UNIX" ];
+          LockPersonality = true;
+          RestrictRealtime = true;
+          RestrictSUIDSGID = true;
+          MemoryDenyWriteExecute = true;
+          SystemCallArchitectures = "native";
+          RuntimeDirectory = "seahub";
+          StateDirectory = "seafile";
+          LogsDirectory = "seafile";
+          ExecStart = ''
+            ${penv}/bin/gunicorn seahub.wsgi:application \
+            --name seahub \
+            --workers ${toString cfg.workers} \
+            --log-level=info \
+            --preload \
+            --timeout=1200 \
+            --limit-request-line=8190 \
+            --bind unix:/run/seahub/gunicorn.sock
+          '';
+        };
+        preStart = ''
+          if [ ! -d "${seahubDir}" ]; then
+              mkdir ${seahubDir}
+              ln -sf ${seahubSettings} ${centralConfigDir}/seahub_settings.py
+              # link media into local seahubDir, for avatars
+              mkdir -p ${seahubDir}/media
+              for m in ${pkgs.python3Packages.seahub}/media/*/ ; do
+                ln -s $m ${seahubDir}/media/
+              done
+              rm ${seahubDir}/media/avatars
+              cp -rf ${pkgs.python3Packages.seahub}/media/avatars ${seahubDir}/media/
+              chmod -R 0755 ${seahubDir}/media/avatars
+              # init database
+              ${pkgs.python3Packages.seahub}/manage.py migrate
+              # create admin account
+              ${pkgs.expect}/bin/expect -c 'spawn ${pkgs.python3Packages.seahub}/manage.py createsuperuser --email=${cfg.adminEmail}; expect "Password: "; send "${cfg.initialAdminPassword}\r"; expect "Password (again): "; send "${cfg.initialAdminPassword}\r"; expect "Superuser created successfully."'
```

Comment:
Should we leave this up to the sysadmin to complete as a task afterwards? We could avoid storing a password in the nix store which will become out of date anyways.

Reply:
This can be left to the admin :+1:, how should we document that?
This "initialPassword" pattern is used by [several](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&query=gitlab.initial) [other](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&query=keycloak.initialAdmin) [services](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&query=moodle.initial): is it seen as a bad practice?

Reply:
Depends who you ask... :eyes: You might consider adding it to the `description` field in `services.seafile.enable`.

Reply:
Since this command expects the whole seafile environment to be set (seaf-server running, seafile conf dir env var, ccnet conf dir, etc) it seems more convenient to keep this here during init.